### PR TITLE
[WIP] Scoreboard changes for 1.8

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR builds off of @games647's changes.

To do:
- Colored display slots
- Fix JavaDoc

### CODE REVIEW

**Comment by SpaceManiac** [comment]
What is the justification for turning criteria into an enum?

**Comment by Aaron1011** [comment]
@SpaceManiac: The underlying strings used by the client to indicate criteria types may change. Making them enum constants abstracts the underlying value from a plugin, and allows the criteria type to be checked at compile-time.

If a string was used instead, an incorrect criteria type would cause either a client-side crash, or be handled silently, with no indication that anything went wrong.

**Comment by SpaceManiac** [comment]
As I understand, invalid criteria (such as `dummy`) are silently ignored by the client. But isn't this solved by the existing `Criterias` set of string constants? Also, there's some use in assigning custom criteria the client ignores but the server (or a plugin) can use to keep track of different events.

What's with the death message visibility? Does Vanilla support this?
Could you also explain the Trigger criteria, and what locking accomplishes?

**Comment by Aaron1011** [comment]
@SpaceManiac: Using string constants would work, but I thought it made more sense for it to be an Enum.

Death message visiblity is part of Vanilla, though it's purely server-side. As the name would suggest, it controls the death messages seen by members of a team. It has exactly the same options as name tag visibility, thus the re-use of the `NametagVisibility` enum.

The [Trigger criteria](http://minecraft.gamepedia.com/Scoreboard) is part of Vanilla, and allows normal, non-op players to modify specific objectives. The `/scoreboard players enable <player> <trigger objective>` command must first be used to enable a trigger objective for a specific player (or `*`) for all players.

Once a player has a trigger objective enabled, players are able to use the [`/trigger`](http://minecraft.gamepedia.com/Commands#trigger) command to either set or add to their score on the objective. After using `/trigger` command, they must be re-enabled for that objective to use it again.

I'm currently working on implementing the `Trigger` objective; I'll push up my changes in a minute.

**Comment by SpaceManiac** [comment]
In my opinion it's simpler to keep criteria as strings as they are now (and there's no need to deal with back-compat shenanigans, which you would need to address). Also note that the page you linked explains there are many criteria which are not, in fact, in that small list... so strings win here.

Looks solid on the rest, though, I think.

**Comment by Aaron1011** [comment]
@SpaceManiac: Okay, I've fixed it.

File: src/main/java/org/bukkit/scoreboard/Scoreboard.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
      * Registers an Objective on this Scoreboard
      *
      * @param name Name of the Objective
-     * @param criteria Criteria for the Objective
+     * @param critera Criteria for the Objective
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/scoreboard/Scoreboard.java]: Typo?

File: src/main/java/org/bukkit/scoreboard/Criteria.java [hunk 2]
```diff
@@ -3,18 +3,20 @@
 /**
  * Criteria names which trigger an objective to be modified by actions in-game
  */
-public class Criterias {
+public class Criteria {
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/scoreboard/Criteria.java]: Changing the name here breaks back-compat, I recommend leaving as is.

File: src/main/java/org/bukkit/scoreboard/Criteria.java [hunk 3]
```diff
@@ -3,18 +3,20 @@
...
         PLAYER_KILLS="playerKillCount";
         TOTAL_KILLS="totalKillCount";
         DEATHS="deathCount";
+        TRIGGER="trigger";
     }
 
-    private Criterias() {}
-}
+    private Criteria() {}
+}
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/scoreboard/Criteria.java]: Missing eof newline.

**Comment by Aaron1011** [comment]
@SpaceManiac: I've made the changes you requested.

File: src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java [hunk 4]
```diff
@@ -100,7 +100,13 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
                 }
                 String name = args[2];
-                String criteria = args[3];
+                String criteria = null;
+                for (String crit: OBJECTIVES_CRITERIA) {
+                    if (args[3].equalsIgnoreCase(crit)) {
+                        criteria = args[3];
+                        break;
+                    }
+                }
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java]: Are you sure on the need for this? Wouldn't it be useful to allow the command to use custom criteria if needed, or the achivement/statistic criteria? Tab-complete can remain as it is currently.

File: src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java [hunk 5]
```diff
@@ -30,8 +30,8 @@
 
     private static final List<String> MAIN_CHOICES = ImmutableList.of("objectives", "players", "teams");
     private static final List<String> OBJECTIVES_CHOICES = ImmutableList.of("list", "add", "remove", "setdisplay");
-    private static final List<String> OBJECTIVES_CRITERIA = ImmutableList.of("health", "playerKillCount", "totalKillCount", "deathCount", "dummy");
-    private static final List<String> PLAYERS_CHOICES = ImmutableList.of("set", "add", "remove", "reset", "list");
+    private static final List<String> OBJECTIVES_CRITERIA = ImmutableList.of("health", "playerKillCount", "totalKillCount", "deathCount", "dummy", "trigger");
+    private static final List<String> PLAYERS_CHOICES = ImmutableList.of("set", "add", "remove", "reset", "list", "enable");
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java]: "test" missing here?

File: src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java [hunk 6]
```diff
@@ -511,7 +614,7 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
                 if (args.length == 2) {
                     return StringUtil.copyPartialMatches(args[1], PLAYERS_CHOICES, new ArrayList<String>());
                 }
-                if (args[1].equalsIgnoreCase("set") || args[1].equalsIgnoreCase("add") || args[1].equalsIgnoreCase("remove")) {
+                if (args[1].equalsIgnoreCase("set") || args[1].equalsIgnoreCase("add") || args[1].equalsIgnoreCase("remove") || args[1].equalsIgnoreCase("enable")) {
```
Comment (SpaceManiac) [inline src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java]: "test" missing here?

**Comment by SpaceManiac** [comment]
My final concern is javadocs (descriptions on classes/methods/params/returns, removal of `@since`). Other than that I'm reasonably certain everything is covered. Thanks for the hard work!

File: src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java (hunk 4 above)
Comment (Aaron1011) [inline src/main/java/org/bukkit/command/defaults/ScoreboardCommand.java]: Yeah, I think so. It shouldn't matter for most users, as they'll just be using the string constants in `Criterias`. I'll change it.

**Comment by Aaron1011** [comment]
@SpaceManiac: I've updated all the JavaDoc.

**Comment by SpaceManiac** [comment]
Merged on the `scoreboard` branch.

**Pull request closed without merging**